src/fold.jl

```diff
@@ -0,0 +1,320 @@
+```julia
+all(rand(folded(Lapace(μ,b),0),1000) .≥ 0) # -> true
+```
+"""
+function folded end
+
+
+#the folded function with no arguments will just return the function itself
+folded(d::ContinuousUnivariateDistribution, ::Nothing, ::Nothing) = d
+
+
+##### Defining the function and the initialization functions #####
+"""
+Generic wrapper for the folded version of a distribution. 
+Holds the original distribution, and two truncated copies of the 
+distribution: one above the crease and one below. 
+"""
+struct Folded{D<:ContinuousUnivariateDistribution, K<:Truncated, S<:Continuous, T <: Real} <: ContinuousUnivariateDistribution
+    original::D      # the original distribution (unfolded)
+    included::K     # The part of the old distribution left unchanged
+    excluded::K     # The folded part that is now part of the new distribution
+    crease::T     # Value at which to make the fold
+    keep_right::Bool # true if distribution folds the left distribution on to the right distribution, false if otherwise
+
+    # Internal constructor
+    function Folded(unfolded::ContinuousUnivariateDistribution, included::K, excluded::K,  crease::T, keep_right::Bool;  check_args=true) where {T <: Real, K <: Truncated}
+        # Check if the crease is within the support of the original distiribution
+        check_args && @check_args(Folded, insupport(unfolded, crease))
+        new{typeof(unfolded), typeof(included), value_support(typeof(unfolded)), T}(unfolded, included, excluded, crease, keep_right)
+    end
+end
+
+# This is the user facing initialization function that will be used in most cases
+"""
+   folded(d::ContinuousUnivariateDistribution, crease::T; keep_right=true) 
+
+Creates a folded distribution from the original distribution `d` by folding the distribution about the `crease`.
+This function defaults to folding the left side onto the right, but by using `keep_right=false` one can reflect the right side onto the left.
+
+"""
+function folded(d::ContinuousUnivariateDistribution, crease::T; keep_right=true) where {T <: Real}
+    lower = minimum(d); 
+    upper = maximum(d);
+    left_distribution = truncated(d, lower, crease);
+    right_distribution = truncated(d, crease, upper);
+    if keep_right
+        included = right_distribution
+        excluded = left_distribution
+    else
+        included = left_distribution
+        excluded = right_distribution
+    end
+    Folded(d, included, excluded, crease, keep_right)
+end
+
+############################################################################
+# This section is there so that typeof(d)(params(d)...) == d works:
+############################################################################
+
+# Define params as the unfolded original parameters, just appended with the crease and keep_right variables
+params(d::Folded) = tuple(params(d.original)..., d.crease, d.keep_right)
+partype(d::Folded) = partype(d.original)
```

Comment:
I'm not sure if this is correct, but `partype` doesn't have a docstring so I'm not sure what it promises. @devmotion ?